### ISSUE TITLE
JavaScript/Typescript Upper_Snake_Case issues

### ISSUE THREAD

**Issue by jkavalec** [highlightjs/highlight.js#4072]
**Describe the issue**
Wen you have non standard code style - using Upper Snake case in this instance - most types of Ids except for variable that is assigned arrow style function, have highlight that is composed of two colors, rendering only first part of Id correctly.

**Which language seems to have the issue?**
TypeScript/JavaScript

**Are you using `highlight` or `highlightAuto`?**
I noticed this issue on StackOverflow and ChatGPT and would like to see naming style I use highlighted properly

**Sample Code to Reproduce**

```typescript
type Only_First_Is_Highlighted = any

const idCamel = "no highlight for basic id"
const snake_camel = "also as expected"
const PascalCase = "here is different Highlight"
const Upper_Snake = "this again breaks"
const SCREAMING_combined = "this is not highlighted as basic id"
const SCREAMING_SNAKE = "this is as ist should be"

const dict = {
	camelKey: "works",
	PascalKey: "different color, but still not broken",
	Upper_Snake_Case: "broken",
}

// everything in this case is consistend id is identified as
// const function and colored same way
const camelCaseFnWorks = () => {}
const snake_works = () => {}
const Upper_Snake_Works = () => {}
const SCREAMING_SNAKE = () => {}

//all below are broken
type Upper_Snake_Case = (min: number, max: number) => number

const Upper_Snake_Case: Upper_Snake_Case = function(min, max) {}

const SCREAMING_SNAKE: Upper_Snake_Case = function(min, max) {}

const Upper_Case = function(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

// lower snake case here is consistent with camel case

const lower_snake_case = function(min, max) {}
const camelCase = function(min, max) {}

const lower_snake_case: Upper_Snake_Case = function(min, max) {}
const camelCase: Upper_Snake_Case = function(min, max) {}
```

Here github code for instance has no broken Upper Snake Case

![image](https://github.com/highlightjs/highlight.js/assets/15125854/2092060a-dd47-4654-940c-711e6a78ff74)

**Expected behavior**
Single Id of any kind should not be colored in 2 separate colors.

**Comment by joshgoebel** [comment]
I think we'd be open to a simple patch to NOT highlight identifiers `Upper_Snake_Works`.  This is no type of idiomatic JS I've ever seen.  If you have a pointer to some specific guides I'd be willing to check them out.  I checked out the GitHub guide just quickly and it had no mention I saw of this time of mix of camel case and snake case. 

It looks like Github is tagging this `smi` and not highligthing it the same as the other variables.  Do we know what `smi` is?

**Comment by joshgoebel** [comment]
>  for this non idiomatic JS.

Exactly. We don't care much about highlighting non idiomatic JS "correctly" or that we poorly highlight non idiomatic JS.  We're solving for the most common use cases here.  However, as I mentioned, if a simple enough patch was proposed to prevent poor highlighting of these mixed Camel+Snake case we'd likely accept it.  Probably a matter of matching on a word boundary in addition to a negative lookahead for the underscore.

We highlight only valid syntax *and* very popular idiomatic patterns that are common.

The highlighting is all based on regex patterns. You could indeed highlight these weirdly named variables/constants if you wrote a custom grammar to do so and wrote the appropriate regex.  If you want to modify grammars or buid your own I'd suggest [reading our docs](https://highlightjs.readthedocs.io/en/latest/) and then becoming familiar with regex patterns for describing matches.

**Comment by jkavalec** [comment]
Ok, I will let you know what I find about it and whether there is some simple solution, I started  go through some code and will look more into how this all works internally, thank you for comment.

**Comment by jkavalec** [comment]
Hi, so I created PR, I discovered one more case where it was working before changes (class declaration).

I looked into the code a little, made a tiny change to one regex, this didn't affect any highlighting that was working before, and all the naming is now consistent across the rules.

I just wasn't sure where to put new markups so I created new one inside TS markups which checks all the things that are common with JS but also has code with TS types.
